srk/src/transitionSystem.ml

```diff
@@ -662,18 +684,168 @@ module Make
     in
     List.map invariants (L.all_loops (L.loop_nest tg))
 
-  let simplify p tg =
+
+  let inline ?(depth=(-1)) tg entry (displayer: t -> unit) (assertion_map : 'x SrkUtil.Int.Map.t) = 
+    let pmap = PHT.create 998 in (* y in pmap[x] means call from x->y *)
+    let qmap = PHT.create 998 in (* (x, z) in qmap[y] means call from x->y via call-edge z *)
+    let greatest = ref (WG.fold_vertex max tg (-1)) in
+    let assertions = ref assertion_map in   
+    PHT.add pmap (entry, -1) PS.empty;
+    PHT.add qmap (entry, -1) PPS.empty;
+    let procedures = 
+      WG.fold_edges (fun (_, w, _) acc -> 
+        match w with 
+        | Call (x, y) -> 
+          begin match PHT.find_opt pmap (x, y) with 
+          | None -> 
+            PHT.add pmap (x, y) PS.empty 
+          | Some _ -> ()
+          end;
+          begin match PHT.find_opt qmap (x, y) with 
+          | None -> 
+            PHT.add qmap (x, y) PPS.empty 
+          | Some _ -> () 
+          end;
+          PS.add (x, y) acc
+        | _ -> acc 
+        ) tg (PS.add (entry, -1) PS.empty) in 
+    let rec dfs (proc: int * int) tg src (f: (vertex * vertex) -> vertex -> vertex * 'a label * vertex -> unit) (visited : ISet.t) = 
+      Printf.printf "dfs: visiting %d\n" src;
+      WG.fold_succ_e (fun (u, w, v) visited -> 
+        f proc src (u, w, v);
+        begin match ISet.find_opt v visited with 
+            | None -> ISet.union (dfs proc tg v f (ISet.add v visited)) visited 
+            | Some _ -> visited 
+            end) tg src visited in 
+      PS.iter (fun (x, y) -> 
+        Printf.printf "populating p/qmaps with dfs... %d %d\n" x y;
+        ignore @@ dfs (x, y) tg x (fun proc src (_, w, v) -> 
+          match w with 
+          | Call (x, y) -> 
+            let caller_set = PHT.find pmap proc in 
+            let callee_set = PHT.find qmap (x, y) in
+            PHT.add pmap proc (PS.add (x, y) caller_set);
+            PHT.add qmap (x, y) (PPS.add (proc, (src, v)) callee_set);
+          | _ -> () 
+          ) (ISet.add x ISet.empty)) procedures; (* populate pmap, qmap *)
+      let compute_sinks () = 
+        PS.fold (fun (x, y) acc -> (* compute sink locations to start inlining from *)
+          match (PHT.find_opt pmap (x, y), PHT.find_opt qmap (x, y)) with 
+          | Some callees, Some callers -> 
+            (* a procedure is considered for inlining if it is (1) a sink in the call graph (2) at least one function calls it.*)
+            if ((PS.cardinal callees) == 0) && ((PPS.cardinal callers) > 0) then PS.add (x, y) acc else 
+              begin 
+                Printf.printf "%d %d is not a sink; num callees = %d num callers = %d\n" x y (PS.cardinal callees) (PPS.cardinal callers);
+                acc
+              end 
+          | (_, _) -> failwith ""
+          ) procedures PS.empty in
+      let copy_subgraph tg src = 
```

Comment:
copy_subgraph doesn't rename variables, so it's unsafe for recursive procedures.  That's ok--complete inlining won't work anyway---but add a comment to the inline function that recursion is not supported, even with the depth bound.